sopel/irc/isupport.py

```diff
@@ -0,0 +1,380 @@
+        if key_ci not in self.__isupport:
+            raise KeyError(key_ci)
+        return self.__isupport[key_ci]
+
+    def __contains__(self, key):
+        return key.upper() in self.__isupport
+
+    def __getattr__(self, name):
+        if name not in self.__isupport:
+            raise AttributeError(name)
+
+        return self.__isupport[name]
+
+    def __setattr__(self, name, value):
+        # make sure you can't set the value of any ISUPPORT attribute yourself
+        if name == '_ISupport__isupport':
+            # allow to set self.__isupport inside of the class
+            super(ISupport, self).__setattr__(name, value)
+        elif name in self.__isupport:
+            # reject any modification of __isupport
+            raise AttributeError("Can't set value for %r" % name)
+        elif name not in self.__dict__:
+            raise AttributeError('Unknown attribute')
+
+    def apply(self, **kwargs):
+        """Build a new instance of :class:`ISupport`.
+
+        :return: a new instance, updated with the latest advertised features
+        :rtype: :class:`ISupport`
+
+        This method applies the latest advertised features from the server:
+        the result contains the new and updated parameters, and doesn't contain
+        the removed parameters (marked by ``-{PARAMNAME}``)::
+
+            >>> updated = {'-AWAYLEN': None, 'NICKLEN': 25, 'CHANNELLEN': 10}
+            >>> new = isupport.apply(**updated)
+            >>> 'CHANNELLEN' in new
+            True
+            >>> 'AWAYLEN' in new
+            False
+
+        """
+        kwargs_ci = dict((key.upper(), value) for key, value in kwargs.items())
+        kept = (
+            (key, value)
+            for key, value in self.__isupport.items()
+            if ('-%s' % key) not in kwargs_ci
+        )
+        updated = dict(itertools.chain(kept, kwargs_ci.items()))
+
+        return self.__class__(**updated)
+
+    @property
+    def CHANLIMIT(self):
+        """Expose ``CHANLIMIT`` as a dict, if advertised by the server.
+
+        This exposes information about the maximum number of channels that the
+        bot can join for each prefix::
+
+            >>> isupport.CHANLIMIT
+            {
+                '#': 70,
+                '&': None,
+            }
+
+        In that example, the bot may join 70 ``#`` channels and any number of
+        ``&`` channels.
+
+        This attribute is not available if the server does not provide the
+        right information, and accessing it will raise an
+        :exc:`AttributeError`.
+
+        .. seealso::
+
+            https://modern.ircdocs.horse/#chanlimit-parameter
+
+        """
+        if 'CHANLIMIT' not in self:
+            raise AttributeError('CHANLIMIT')
+
+        return dict(self['CHANLIMIT'])
+
+    @property
+    def CHANMODES(self):
+        """Expose ``CHANMODES`` as a dict, if advertised by the server.
+
+        This exposes information about 4 types of channel::
```

Comment:
```suggestion
        This exposes information about 4 types of channel modes::
```

Also, what about the allowed additional types that may follow in a tuple at position 5? Are those only to be accessible through subscripting the `ISupport` instance?

Reply:
Yup. `other_chantype_modes = isupports['chanmodes'][5]` where `other_chantype_modes` is now a tuple of everything else.

Reply:
I'm not sure I like `isupport['CHANMODES']` to return different information than `isupport.CHANMODES`. Shouldn't they present the same data, with the only difference being how it was obtained?

Reply:
See `isupport['key']`  as a low level API with raw data, and `isupport.SPECIFIC_KEY` as a high level API with structured and documented data.

Since the modern ircdoc tend to discourage from using extra types:

> However, server authors SHOULD NOT extend this parameter without good reason, and SHOULD CONSIDER whether their mode would work as one of the existing types instead.
> **The behaviour of any additional types is undefined.**

And since the behaviour of additional types is undefined, I'd prefer not to expose them in the high level API.